src/lib/homestore/shard_manager.cpp

```diff
@@ -1,15 +1,276 @@
+    if (msg_header->payload_crc != crc) {
+        LOGWARN("replication message body is corrupted with crc error, lsn:{}", lsn);
+	return false;
+    }
+    *msg = std::move(shard_msg);
+    return true;
+}
+
+void HSHomeObject::on_pre_commit_shard_msg(int64_t lsn, sisl::blob const& header, sisl::blob const& key,
+                                           void* user_ctx) {
+    std::string shard_msg;
+    if (!precheck_and_decode_shard_msg(lsn, header, key, &shard_msg)) {
+        //header is broken, nothing to do;
+        return;
+    }
+
+    auto shard = deserialize_shard(shard_msg);
+    if (shard.info.state != ShardInfo::State::OPEN) {
+        //it is not an create shard msg, nothing to do;
+        return;
+    }
+
+    //write shard header to ReplDev which will bind the newly created shard to one underlay homestore physical chunk;
+    using namespace homestore;
+    auto datasvc_page_size = HomeStore::instance()->data_service().get_page_size();
+    if (shard_msg.size() % datasvc_page_size != 0) {
+        uint32_t need_size = (shard_msg.size() / datasvc_page_size + 1)  * datasvc_page_size;
+        shard_msg.resize(need_size);
+    }
+
+    auto write_buf = iomanager.iobuf_alloc(512, shard_msg.size());
+    std::memcpy(r_cast<uint8_t*>(write_buf), r_cast<const uint8_t*>(shard_msg.c_str()), shard_msg.size());
+
+    sisl::sg_list sgs;
+    sgs.size = shard_msg.size();
+    sgs.iovs.emplace_back(iovec(static_cast<void*>(write_buf), shard_msg.size()));
+    std::vector< homestore::BlkId > out_blkids;
+    auto future = HomeStore::instance()->data_service().async_alloc_write(sgs, homestore::blk_alloc_hints(), out_blkids);
+    bool write_result = std::move(future).get();
+    iomanager.iobuf_free(write_buf);
+    if (!write_result) {
+        LOGWARN("write lsn {} msg to homestore data service is failed", lsn);
+        return;
+    }
+    RELEASE_ASSERT(out_blkids.size() == 1, "output blkids num is not 1");
+    shard.chunk_id = out_blkids[0].get_chunk_num();
+    std::scoped_lock lock_guard(_flying_shard_lock);
+    auto [_, happened] = _flying_shards.emplace(lsn, std::move(shard));
+    RELEASE_ASSERT(happened, "duplicated flying create shard msg");
+}
+
+void HSHomeObject::on_rollback_shard_msg(int64_t lsn, sisl::blob const& header, sisl::blob const& key,
+                                         void* user_ctx) {
+    std::string shard_msg;
+    if (!precheck_and_decode_shard_msg(lsn, header, key, &shard_msg)) {
+        //header is broken, nothing to do;
+        return;
+    }
+
+    auto shard = deserialize_shard(shard_msg);
+    if (shard.info.state != ShardInfo::State::OPEN) {
+        //it is not an create shard msg, nothing to do;
+        return;
+    }
+    
+    std::scoped_lock lock_guard(_flying_shard_lock);
+    auto iter = _flying_shards.find(lsn);
+    if (iter != _flying_shards.end()) {
+        _flying_shards.erase(iter);
+    }
```

Comment:
No, I think.  I think the written shard header will be used for recovery in case of meta blks are lost from the previous discussion meeting.

Reply:
if we rollback the `pre_commit`, this means this is a canceled `create_shard`, why to recovery this?

coming to this, i have another question. we rolling back a `pre_commit`, seems nuraft does not append a log for rollback, what the statemachine does is just call the `rollback` function(i just only read the nuraft doc, but not deep dive the code for now, so if this is wrong, please correct me). so , if it crashes when we called `pre_commit` but before or not finish calling `statemachin#rollback` , after restart who will take the task to finish the rollback job

Reply:
1. yes, it is absolutely more clear to truncate/overwrite the shard header which is already written in on_pre_commit(), but homestore DataService is append-only model, I think overwritten is conflict with this model.  even this shard header is leaked,  I think it will be finalized recycled by chunk compaction, right? because this log is rollbacked and will never be commited, so from the homeobject side,  it can not see this shard any more(whether now or after recovery).
2. I think your mentioned case may happens,  restart happens after on_pre_commit() but before on_rollback(), in such case, this shard header is leaked, but same to question 1, it will be recycled finalized.